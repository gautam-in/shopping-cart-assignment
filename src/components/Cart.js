import React, { Component } from "react";
import CartItem from "./CartItem";
import axios from "axios";
import { Link } from "react-router-dom";

class Cart extends Component {
  constructor(props) {
    super(props);

    this.state = {
      productList: [],
    };
  }

  componentDidMount() {
    this.productListData();
  }

  productListData = () => {
    axios.get(`http://localhost:5000/products`).then((res) => {
      if (res && res.status === 200) {
        const response = res.data;
        this.setState({ productList: response });
      } else {
        this.setState({
          productList: [],
        });
      }
    });
  };

  handleButtonClick = () => {
    this.props.history.push("/home");
  };

  render() {
    return (
      <>
        {sessionStorage.getItem("currentUser") ? (
          <section className="cart">
            {this.props.mainCartItems.length > 0 ? (
              <article className="cart-filled">
                <h2>My Cart</h2>
                <section>
                  {this.props.mainCartItems.map((_) => {
                    return (
                      <main key={_.id} className="item">
                        <CartItem
                          productId={_.id}
                          productSingle={_}
                          productList={this.state.productList}
                          updateCartItemCount={this.props.updateCartItemCount}
                        />
                      </main>
                    );
                  })}
                  <figure className="advertisement">
                    <img
                      className="addv-image"
                      src="../static/images/lowest-price.png"
                      alt="Lowest price gauranteed"
                    />{" "}
                    you wont find it cheaper anywhere!
                  </figure>
                </section>
              </article>
            ) : (
              <article className="cart-empty">
                <div>
                  <h2>No items in your cart</h2>
                  <p style={{ textAlign: "center" }}>
                    Your favourite items are just a click away
                  </p>
                </div>
              </article>
            )}
            <article className="buttons">
              <button
                type="button"
                onClick={this.handleButtonClick}
                onKeyPress={this.handleButtonClick}
                tabIndex={0}
                style={{ zIndex: 100 }}
              >
                {this.props.mainCartItems.length > 0
                  ? `Proceed to checkout`
                  : "Start Shopping"}
              </button>
            </article>
          </section>
        ) : (
          <section className="doLogin">
            <h2
              style={{
                display: "flex",
                justifyContent: "center",
                alignItems: "center",
                height: "70vh",
                flexDirection: "column",
              }}
            >
              <Link to="/signin">Please Click here to Login</Link>
              <h4>
                <Link
                  to="/signup"
                  style={{
                    color: "red",
                    letterSpacing: "1px",
                  }}
                >
                  If you don't have an account please create!
                </Link>
              </h4>
            </h2>
          </section>
        )}
      </>
    );
  }
}

export default Cart;
