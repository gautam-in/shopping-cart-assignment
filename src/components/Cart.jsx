import React from "react";

const Cart = ({
  products,
  totalPrices,
  addToCartHandler,
  removeToCartHandler,
  closeCartHandler,
}) => {
  return (
    <div
      className="relative  md:hidden h-full w-full z-10"
      aria-labelledby="slide-over-title"
      role="dialog"
      aria-modal="true"
    >
      <div className="fixed  right-0 flex max-w-full top-12 sm:top-20 inset-y-0">
        <div className="w-screen ">
          <div className="flex h-full flex-col bg-light shadow-xl">
            <div className="flex-1 overflow-y-auto  ">
              <div className=" flex flex-col h-full w-full bg-light ">
                <div className="h-2"></div>
                <div className="  h-12 mt-5 pl-6 flex items-center  bg-white">
                  <label className="text-2xl  font-bold">
                    My Cart
                    {products?.length > 0
                      ? "(items" + products?.length + ")"
                      : ""}
                  </label>
                </div>
                <div className="h-3"></div>
                <div className="flex flex-col flex-grow bg-white">
                  {products?.map((item, index) => {
                    return (
                      <React.Fragment key={item.id + "div2"}>
                        <div className="flex flex-row  p-3">
                          <img
                            height={70}
                            width={70}
                            src={item.imageURL}
                            alt="/"
                          ></img>
                          <div className="flex pl-2 flex-grow flex-col">
                            <p className="text-sm font-semibold">{item.name}</p>
                            <div className="flex items-center flex-grow">
                              <button
                                onClick={() => {
                                  addToCartHandler(item.id);
                                }}
                                className="h-7 w-7 bg-rose-500 text-white"
                              >
                                +
                              </button>
                              <label className="ml-2">{item.qty}</label>
                              <button
                                onClick={() => {
                                  removeToCartHandler(item.id, item.qty);
                                }}
                                className="h-7 w-7 bg-rose-500 ml-2 mr-2 text-white"
                              >
                                -
                              </button>
                              <label className="text-xl mr-2">x</label>
                              <label className="text-lg">
                                <span className="text-sm"> Rs.</span>
                                {item.price}
                              </label>
                              <label className="ml-auto font-medium text-lg">
                                <span className="text-sm"> Rs.</span>
                                {item.price * item.qty}
                              </label>
                            </div>
                          </div>
                        </div>
                        <div className="h-3 bg-light"></div>
                      </React.Fragment>
                    );
                  })}
                  <div className="flex-grow flex w-full items-center justify-center  bg-light">
                    {products?.length === 0 && (
                      <div className=" ">
                        <p className="text-center font-bold">
                          No items in your cart.
                        </p>
                        <p className="text-center">
                          Your favourite items are just a click away
                        </p>
                      </div>
                    )}
                  </div>
                  <div
                    className="mt-auto    flex flex-col items-center justify-center"
                    style={{
                      boxShadow: "0px -1px 2px rgba(0, 0, 0, 0.3)",
                    }}
                  >
                    {products?.length > 0 && (
                      <label className="text-sm p-2">
                        promo code can be applied on payment page
                      </label>
                    )}
                    {products?.length > 0 && (
                      <button
                        className="bg-rose-600 h-9  text-white w-full "
                        onClick={closeCartHandler}
                      >
                        <div className="flex flex-row cursor-pointer">
                          <label className="pl-5 cursor-pointer">
                            Proceed to Checkout
                          </label>
                          <label className="ml-auto pr-5 cursor-pointer">
                            {totalPrices ? totalPrices : 0}
                          </label>
                        </div>
                      </button>
                    )}
                    {products?.length === 0 && (
                      <button
                        className="bg-rose-600 h-9  text-white w-full "
                        onClick={closeCartHandler}
                      >
                        Start Shopping
                      </button>
                    )}
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Cart;
