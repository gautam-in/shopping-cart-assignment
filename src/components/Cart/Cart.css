.Cart {
  background-color: white;
  padding: 8px;
  height: 100%;
  box-sizing: border-box;
  background: lightgrey;
  display: flex;
  width: 8rem;
  justify-content: space-evenly;
  cursor: pointer;
}

.Cart img {
  height: 25px;
  width: auto;
  filter: invert(27%) sepia(51%) saturate(2878%) hue-rotate(346deg)
    brightness(104%) contrast(97%);
}
