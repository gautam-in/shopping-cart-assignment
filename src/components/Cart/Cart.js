import React from "react";
import { connect } from "react-redux";

import axios from "../../axios-orders";
import withErrorHandler from "../../hoc/withErrorHandler/withErrorHandler";
import * as actions from "../../store/actions/index";
// import cartlogo from "../../assets/images/cart.svg";
import classes from "./Cart.css";

const Cart = (props) => {
  return (
    <div className={classes.Cart} onClick={() => props.onCartClick(true)}>
      <img src={`../../static/images/cart.svg`} alt="Cart img" />
      <span>{props.cartList.length} items</span>
    </div >
  );
};
const mapStateToProps = (state) => {
  return {
    cartList: state.product.addToCartProducts,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    onCartClick: (value) => dispatch(actions.cartHandler(value)),
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(withErrorHandler(Cart, axios));
