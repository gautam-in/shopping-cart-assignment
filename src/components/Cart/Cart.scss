.no-item-text {
  text-align: center;
}
.cart-main-cont {
  position: absolute;
  display: none;
  right: 0;
  top: 65px;
  background: #efefef;
  z-index: 3;
  .cart-container {
    position: relative;
    height: 81.5vh;
    width: 35vw;
    .checkout {
      position: absolute;
      padding: 15px;
      left: 0;
      right: 0;
      bottom: 0;
      background: #fff;
    }
    .checkout-button {
      background: #e83583;
      border: none;
      width: 100%;
      color: #fff;
      font-size: 1rem;
      padding: 15px;
    }
    .checkout-heading {
      margin: 0px 0px 15px 0px;
      text-align: center;
    }
    .header-container {
      background: #000;
      display: flex;
      justify-content: space-between;
      padding: 0px 15px;
      .heading {
        color: #fff;
      }
    }
    .cart_list {
      list-style: none;
      padding: 0px;
      margin-left: 0px;
      max-height: 62vh;
      overflow: auto;
      .quantity {
        font-size: 1rem;
      }
    }
    .cart_list-details {
      width: 100%;
    }
    .cart_list-item {
      background: #fff;
      display: flex;
      align-items: flex-start;
      margin-bottom: 10px;
    }
    .cart_list-item-total {
      font-size: 1rem;
      margin: 0px 15px 0px 0px;
      float: right;
    }
    .cart_list-price {
      font-size: 1rem;
      padding-left: 10px;
    }
    .cart_list-img {
      width: 80px;
      padding: 10px;
    }
    .cart_list-name {
      font-size: 1.1rem;
      font-weight: 600;
      margin: 10px 0px;
      text-align: left;
    }
  }
}
.btn-close {
  background-color: black;
  border: black;
  &:after {
    position: absolute;
    -webkit-transform: translate(-50%, -50%);
    -moz-transform: translate(-50%, -50%);
    -ms-transform: translate(-50%, -50%);
    -o-transform: translate(-50%, -50%);
    transform: translate(-50%, -50%);
    color: #fff;
    font-size: 50px;
    content: "\d7";
    top: 40px;
    right: 18px;
    cursor: pointer;
  }
}
.btn-increment {
  background-color: #e83583;
  border: none;
  padding: 2px 5px 2px 5px;
  height: 30px;
  width: 30px;
  border-radius: 50%;
  margin: 0px 15px;
  font-size: 22px;
  font-weight: bold;
  color: #fff;
  cursor: pointer;
}
.btn-decrement {
  background-color: #e83583;
  border: none;
  padding: 2px 5px 2px 5px;
  height: 30px;
  width: 30px;
  border-radius: 50%;
  margin: 0px 15px;
  font-size: 22px;
  font-weight: bold;
  cursor: pointer;
  color: #fff;
}
@media only screen and (min-device-width: 481px) and (max-device-width: 1024px) {
  .cart-main-cont {
    position: fixed;
    .cart-container {
      width: 100vw;
      height: 95vh;
    }
  }
}
@media only screen and (min-device-width: 320px) and (max-device-width: 480px) {
  .cart-main-cont {
    position: fixed;
    display: none;
    right: 0;
    top: 10vh;
    bottom: 0;
    left: 0;
    background: #efefef;
    z-index: 3;
    .cart-container {
      position: relative;
      width: 100vw;
      height: 90vh;
      .checkout {
        position: absolute;
        padding: 15px;
        left: 0;
        right: 0;
        bottom: 0;
        background: #fff;
      }
      .checkout-button {
        background: #e83583;
        border: none;
        width: 100%;
        color: #fff;
        font-size: 1rem;
        padding: 15px;
      }
      .checkout-heading {
        margin: 0px 0px 15px 0px;
        text-align: center;
        font-size: 1rem;
      }
      .header-container {
        background: #000;
        display: flex;
        justify-content: space-between;
        padding: 0px 15px;
        .heading {
          color: #fff;
          font-size: 1.6rem;
        }
      }
      .cart_list {
        max-height: 71vh;
        overflow: auto;
        .quantity {
          font-size: 1rem;
          padding: 0px 15px;
        }
        .into-qty {
          font-size: 1rem;
          padding: 0px 5px;
        }
      }
      .cart_list-item {
        background: #fff;
        display: flex;
        align-items: flex-start;
        margin-bottom: 10px;
      }
      .cart_list-item-total {
        font-size: 1rem;
      }
      .cart_list-price {
        font-size: 1rem;
        padding: 0px 15px;
        padding-left: 10px;
      }
      .cart_list-img {
        padding: 10px;
      }
      .cart_list-name {
        font-size: 1rem;
        font-weight: 600;
        margin: 10px 0px;
        text-align: left;
        display: block;
      }
    }
  }
}
