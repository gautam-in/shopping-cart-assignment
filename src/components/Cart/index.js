import React from "react";
import { useSelector, useDispatch } from "react-redux";

import "./style.scss";
import { updateCart, removeCart } from "../../store/action";

export default function Cart({ closeCart, ...props }) {
  const { item, cart } = useSelector((state) => state);
  const dispatch = useDispatch();

  return (
    <div className="cart-wrapper">
      <div className="header-container">
        <h1 className="heading">
          My cart {item > 0 ? `( ${item} Item )` : null}
        </h1>
        <button className="btn-close" onClick={() => closeCart(false)} />
      </div>

      <ul className="cart_list">
        {cart.length > 0 ? (
          cart.map((item) => (
            <li className="cart_list-item" key={item.id}>
              <img
                src={item.imageURL}
                className="cart_list-img"
                alt={item.name}
              />
              <div className="cart_list-details">
                <h2 className="cart_list-name truncate">{item.name}</h2>
                <button
                  className="btn-increment"
                  onClick={() => {
                    dispatch(updateCart(item));
                  }}
                >
                  +
                </button>
                <span className="quantity">{item.qty}</span>
                <button
                  className="btn-decrement"
                  onClick={() => {
                    dispatch(removeCart(item));
                  }}
                >
                  -
                </button>
                <span className="into-qty">X</span>
                <span className="cart_list-price">Rs. {item.price}</span>
                <span className="cart_list-item-total">
                  Rs.{item.qty * item.price}
                </span>
              </div>
            </li>
          ))
        ) : (
          <div style={{ textAlign: "center" }}>
            <b>No items in your cart</b>
            <p>Your favourite items are just a click away</p>
          </div>
        )}
      </ul>

      <div className="checkout">
        <p className="checkout-heading">
          Promo code can be applied on the checkout page.
        </p>
        {item > 0 ? (
          <button
            type="button"
            name="button"
            className="btn-cart-checkout checkout-button"
            onClick={() => closeCart(false)}
          >
            <span className="checkout-text">Proceed to checkout</span>
            <span className="checkout-price">
              {" "}
              Rs.{" "}
              {cart.reduce(function (total, item) {
                return total + item.price * item.qty;
              }, 0)}
            </span>
          </button>
        ) : (
          <button
            type="button"
            name="button"
            className="btn-cart-checkout checkout-button"
            onClick={() => closeCart(false)}
          >
            <span className="checkout-text">Start Shopping</span>
          </button>
        )}
      </div>
    </div>
  );
}
