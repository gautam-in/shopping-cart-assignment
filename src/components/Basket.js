import React from "react";
import {connect} from 'react-redux';
import {resetCartData} from '../actions/productActions';
import Cart from '../presentations/Cart';
import toastr from 'toastr';
import Header from '../presentations/Header';
import Constants from '../constants';
class Basket extends React.Component{
    constructor(props){
        super(props);
        this.state={}
        // this.checkoutComplete = this.checkoutComplete.bind(this);
    }
    checkoutComplete = () => {
        if(this.props.productInfo.cartItems && this.props.productInfo.cartItems.length){
            toastr.success(Constants.TEXTS.TOASTS.cart_emptied,Constants.TEXTS.TOASTS.order_placed,{timeOut:1000});
        }
        this.props.resetCartData();
        this.props.router.push('/');
    }
    showCartView=()=>{
       
    }
    render(){
    const {productInfo,router} = this.props;
    return(<div className="basket-page">
        <Header router={router} cartClick={this.showCartView} cartInfo={productInfo.cartItems} />
        <Cart hideHeader={true} cartData={productInfo.cartItems} checkoutComplete={this.checkoutComplete} />
        </div>)
    }
}
function mapStateToProps(state){
    return{
        productInfo: state.productReducer
    }
}
const mapDispatchToProps = (dispatch,ownProps) => {
    return {
        resetCartData: ()=>{dispatch(resetCartData())}
    }
}
export default connect(mapStateToProps,mapDispatchToProps)(Basket);