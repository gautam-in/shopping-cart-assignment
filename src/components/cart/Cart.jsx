import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  addItemsToCart,
  removeItemsFromCart,
} from "../../redux/cart/actionCreator";
import "./Cart.css";

function Cart(props) {
  const cartData = useSelector((state) => state.getCartDetails.cartItems);
  const dispatch = useDispatch();

  useEffect(() => {
    document.body.style.overflow = "hidden";
    return () => (document.body.style.overflow = "unset");
  }, []);

  const addItem = (productData) => {
    let prodObj = {
      id: productData.id,
    };
    dispatch(addItemsToCart(prodObj));
  };

  const removeItem = (productData) => {
    let prodObj = {
      id: productData.id,
    };
    dispatch(removeItemsFromCart(prodObj));
  };

  return (
    <div className="app-cart">
      <div className="cart-section">
        <div className="cart-header">
          <p>My Cart ({cartData.length} item)</p>
          <button onClick={() => props.openCart(true)} aria-label="close cart">
            X
          </button>
        </div>
        <ul className="cart-information">
          {cartData &&
            cartData.map((product, index) => (
              <li className="product-details" key={product.id} tabIndex="0">
                <img
                  src={product.image}
                  alt={product.name}
                  loading="lazy"
                  width="45"
                ></img>
                <div className="product-info">
                  <p>{product.name}</p>
                  <div className="quantity">
                    <button
                      className="cart-btn"
                      onClick={removeItem.bind(null, product)}
                    >
                      <span aria-label="remove item">-</span>
                    </button>
                    <span>{product.count}</span>
                    <button
                      className="cart-btn"
                      onClick={addItem.bind(null, product)}
                    >
                      <span aria-label="add Item">+</span>
                    </button>
                    <span>X</span>
                    <span>{product.price}</span>
                  </div>
                </div>
                <p>
                  <b>{product.price * product.count}</b>
                </p>
              </li>
            ))}

          {cartData.length ? (
            <li className="low-price" tabIndex="0">
              <img
                src="/static/images/lowest-price.png"
                alt="lowest Price Logo"
                loading="lazy"
                width="100"
              ></img>
              <p>You Won't find it cheaper anywhere</p>
            </li>
          ) : (
            <li className="empty-cart centre" tabIndex="0">
              <p>
                <b>No Items in your cart</b>
              </p>
              <p>Your favourite items are just a click away</p>
            </li>
          )}
        </ul>
        <div className="cart-footer centre">
          {cartData.length ? (
            <>
              <p>Promo Code can be applied to payment page</p>
              <button className="app-btn">
                <span>Proceed to Checkout</span>
                <span>
                  Rs :
                  {Object.values(cartData).reduce(
                    (t, { price, count }) => t + price * count,
                    0
                  )}
                </span>
              </button>
            </>
          ) : (
            <button className="app-btn centre">Start Shopping</button>
          )}
        </div>
      </div>
    </div>
  );
}

export default React.memo(Cart);