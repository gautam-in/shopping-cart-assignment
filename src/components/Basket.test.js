import React from 'react';
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import Basket from './Basket';
import Cart from '../presentations/Cart';
const buildStore = configureStore([thunk]);

Enzyme.configure({adapter: new Adapter(),disableLifecycleMethods: false,lifecycleExperimental: true,});
let wrapper; let shallowWrapper; let store;
const props = {productReducer:{cartItems:[]},resetCartData:()=>{},signinReducer:{userData:{userEmail:''}},signUpReducer:{userData:{userEmail:''}}};
describe('Basket',()=>{
    beforeAll(()=>{
        store = buildStore(props);
        shallowWrapper = shallow(<Provider store={store} {...props}>
            <Basket />
        </Provider>);
        props.router = shallowWrapper.instance().props.router;
        store = buildStore(props);
    });
    beforeEach(()=>{
        wrapper = mount(<Provider store={store} {...props}>
            <Basket />
        </Provider>);
    });
    afterEach(()=>{
        wrapper.unmount();
    });
    afterAll(()=>{
        shallowWrapper.unmount();
    });
    it('Rendered Basket Correctly',()=>{
        expect(shallowWrapper).toMatchSnapshot()
    });
    describe('initial load',()=>{
        it('check for header',()=>{
            expect(wrapper.find('Header').exists()).toBe(true);
        });
        it('check for footer',()=>{
            expect(wrapper.find('Footer').exists()).toBeFalsy();
        });
        describe('basket state',()=>{
            it('inital state of basket',()=>{
                expect(shallowWrapper.dive().dive().state()).toEqual({});
            })
        })
    });
    it('cart component present',()=>{
        expect(wrapper.find('Cart').exists()).toBeTruthy();
    })
})