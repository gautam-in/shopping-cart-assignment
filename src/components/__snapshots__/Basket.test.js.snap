// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket Rendered Basket Correctly 1`] = `ShallowWrapper {}`;
